src/simulators/statevector/qubitvector.hpp

```diff
@@ -1801,150 +1709,102 @@ void QubitVector<data_t>::apply_diagonal_matrix(const uint_t qubit,
-    for (uint_t i = 0; i < dst_dim; ++i) {
-      if (filled[i])
-        continue;
-
-      //  3. allocate vmat to u based on u(i, i) and delta
-      u[i                             + (i + 0)                      * dst_dim] = sorted_vmat[0 + 0 * src_dim];
-      u[i                             + (i + low_delta)              * dst_dim] = sorted_vmat[0 + 1 * src_dim];
-      u[i                             + (i + high_delta)             * dst_dim] = sorted_vmat[0 + 2 * src_dim];
-      u[i                             + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[0 + 3 * src_dim];
-      u[(i + low_delta)               + (i + 0)                      * dst_dim] = sorted_vmat[1 + 0 * src_dim];
-      u[(i + low_delta)               + (i + low_delta)              * dst_dim] = sorted_vmat[1 + 1 * src_dim];
-      u[(i + low_delta)               + (i + high_delta)             * dst_dim] = sorted_vmat[1 + 2 * src_dim];
-      u[(i + low_delta)               + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[1 + 3 * src_dim];
-      u[(i + high_delta)              + (i + 0)                      * dst_dim] = sorted_vmat[2 + 0 * src_dim];
-      u[(i + high_delta)              + (i + low_delta)              * dst_dim] = sorted_vmat[2 + 1 * src_dim];
-      u[(i + high_delta)              + (i + high_delta)             * dst_dim] = sorted_vmat[2 + 2 * src_dim];
-      u[(i + high_delta)              + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[2 + 3 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + 0)                      * dst_dim] = sorted_vmat[3 + 0 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + low_delta)              * dst_dim] = sorted_vmat[3 + 1 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + high_delta)             * dst_dim] = sorted_vmat[3 + 2 * src_dim];
-      u[(i + low_delta + high_delta)  + (i + low_delta + high_delta) * dst_dim] = sorted_vmat[3 + 3 * src_dim];
-
-      filled[i] = true;
-      filled[i + low_delta] = true;
-      filled[i + high_delta] = true;
-      filled[i + low_delta + high_delta] = true;
-    }
-    //TODO: } else if (src_qubits.size() == 3) {
+  if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
+   && mat[1] == 0.0 && mat[5] == 0.0 && mat[9]  == 0.0
+   && mat[2] == 0.0 && mat[6] == 0.0                   && mat[14] == 0.0
+   && mat[3] == 0.0 && mat[7] == 1.0 && mat[11] == 0.0 && mat[15] == 0.0
+   ) {
+    // Lambda function for 2-qubit u1-cx matrix multiplication
+    auto lambda = [&](const areg_t<4> &inds, const cvector_t &_mat)->void {
+      std::array<complex_t, 4> cache;
+      for (size_t i = 1; i < 4; i++)
+        cache[i] = data_[inds[i]];
+      data_[inds[1]] = _mat[13] * cache[3];
+      data_[inds[2]] = _mat[10] * cache[2];
+      data_[inds[3]] = cache[1];
+    };
+    apply_lambda(lambda, areg_t<2>( { { qubit0, qubit1 } }), mat);
+  } else if (mat[0] == 1.0 && mat[4] == 0.0 && mat[8]  == 0.0 && mat[12] == 0.0
```

Comment:
Same as above here

Reply:
I removed specialization for 2qubit matrix operation. diagonal matrix operation covers all.